driver/registry_memory.go

```diff
@@ -101,6 +111,108 @@ func (r *RegistryMemory) RuleMatcher() rule.Matcher {
 	return r.ruleRepository
 }
 
+func (r *RegistryMemory) isCachedUpstreamTransportDirty(certFile string) (bool, error) {
+
+	// No cached transport yet, but configuration requires it, so handle as if it was dirty so we get one.
+	if r.upstreamTransport == nil {
+		return true, nil
+	}
+
+	// The transport is dirty if the cert file path changed.
+	if certFile != r.cachedCertFilePath {
+		return true, nil
+	}
+
+	// Transport is dirty if the certificate bytes changed or the modification time,
+	// but only check this based on the ca_refresh_frequency option to reduce IO impact.
+	caRefreshFrequency := r.c.ProxyServeUpstreamCaRefreshFrequency()
+	if caRefreshFrequency <= 0 {
+		return false, nil // Periodic refresh of the Ca is diabled.
+	}
+
+	if r.upstreamRequestCount < int64(caRefreshFrequency) {
+		return false, nil
+	}
+	r.upstreamRequestCount = 0
+
+	stat, err := os.Stat(certFile)
+	if err != nil {
+		return false, err
+	}
+
+	if stat.Size() != (*r.cachedCertFileStat).Size() || stat.ModTime() != (*r.cachedCertFileStat).ModTime() {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func createUpstreamTransportWithCertificate(certFile string) (*http.Transport, *os.FileInfo, error) {
+	transport := &(*http.DefaultTransport.(*http.Transport)) // shallow copy
+
+	// Get the SystemCertPool or continue with an empty pool on error
+	rootCAs, err := x509.SystemCertPool()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	stat, err := os.Stat(certFile)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	certs, err := ioutil.ReadFile(certFile)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Append our cert to the system pool
+	if ok := rootCAs.AppendCertsFromPEM(certs); !ok {
+		return nil, nil, errors.New("No certs appended, only system certs present, did you specify the correct cert file?")
+	}
+
+	transport.TLSClientConfig = &tls.Config{
+		InsecureSkipVerify: false,
+		RootCAs:            rootCAs,
+	}
+
+	return transport, &stat, nil
+}
+
+// UpstreamTransport decides the transport to use for the upstream for the request.
+func (r *RegistryMemory) UpstreamTransport(req *http.Request) (http.RoundTripper, error) {
+
+	r.upstreamRequestCount++
```

Comment:
Since we probably have highly concurrent system this might cause issues. I think it would make sense to use atomic.Uint instead ( https://golang.org/pkg/sync/atomic/ ) to count up here :)